File: streaming-end-to-end/python/streaming_demo.py
"""
Streaming end-to-end demo — Python (OpenAI SDK).

Three patterns:
1. Basic streaming with content_filter handling.
2. json_schema streaming — buffer-then-parse the accumulated content.
3. Reconnect wrapper that retries on connection-class errors only.

Usage:
    pip install -r requirements.txt
    export DVARA_URL="http://localhost:8080/v1"
    export DVARA_API_KEY="gw_<your-dvara-key>"
    python streaming_demo.py

See https://dvarahq.com/docs/cookbook/streaming-end-to-end for the
narrated walkthrough.
"""

import json
import os
import time
from typing import Iterable

from openai import APIConnectionError, OpenAI

DVARA_URL = os.getenv("DVARA_URL", "http://localhost:8080/v1")
DVARA_API_KEY = os.getenv("DVARA_API_KEY", "your-dvara-api-key")

client = OpenAI(base_url=DVARA_URL, api_key=DVARA_API_KEY)


def basic_streaming() -> None:
    """Print each delta as it arrives. Treat content_filter as graceful end."""
    print("=== 1. Basic streaming ===")
    stream = client.chat.completions.create(
        model="gpt-4o",
        messages=[{"role": "user", "content": "Write a haiku about programming."}],
        stream=True,
    )
    for chunk in stream:
        delta = chunk.choices[0].delta
        if delta.content:
            print(delta.content, end="", flush=True)
        if chunk.choices[0].finish_reason == "content_filter":
            print("\n[stream stopped: governance enforcement]", end="")
    print("\n")


def json_schema_streaming() -> None:
    """Stream a structured-output response and parse once complete.

    The gateway flattens any provider-side tool-use rewrite (Anthropic,
    Bedrock) into normal content deltas, so this code works no matter
    which provider the route picked.
    """
    print("=== 2. json_schema streaming ===")
    schema = {
        "type": "object",
        "properties": {
            "name": {"type": "string"},
            "age": {"type": "integer"},
        },
        "required": ["name", "age"],
    }
    stream = client.chat.completions.create(
        model="gpt-4o",
        messages=[{"role": "user", "content": "Extract: John is 30 years old."}],
        response_format={
            "type": "json_schema",
            "json_schema": {"name": "person", "schema": schema, "strict": True},
        },
        stream=True,
    )
    buffer: list[str] = []
    for chunk in stream:
        if chunk.choices[0].delta.content:
            buffer.append(chunk.choices[0].delta.content)

    result = json.loads("".join(buffer))
    print(f"name: {result['name']}, age: {result['age']}\n")


def stream_with_retry(messages, attempts: int = 3, base_backoff: float = 0.5) -> Iterable:
    """Yield chunks, retrying with exponential backoff on connection errors.

    Crucially, this wrapper only retries on transport-class failures.
    A finish_reason: content_filter is an enforcement decision — the
    same content would trigger the same rule on retry, so don't loop.
    """
    for attempt in range(attempts):
        try:
            for chunk in client.chat.completions.create(
                model="gpt-4o", messages=messages, stream=True,
            ):
                yield chunk
            return
        except APIConnectionError:
            if attempt == attempts - 1:
                raise
            time.sleep(base_backoff * (2 ** attempt))


def reconnect_demo() -> None:
    print("=== 3. Reconnect wrapper ===")
    for chunk in stream_with_retry([{"role": "user", "content": "Say hello briefly."}]):
        delta = chunk.choices[0].delta
        if delta.content:
            print(delta.content, end="", flush=True)
    print("\n")


if __name__ == "__main__":
    basic_streaming()
    json_schema_streaming()
    reconnect_demo()

File: streaming-end-to-end/nodejs/streaming_demo.mjs
/**
 * Streaming end-to-end demo — Node.js (OpenAI SDK).
 *
 * Three patterns:
 *   1. Basic streaming with content_filter handling.
 *   2. json_schema streaming — buffer-then-parse the accumulated content.
 *   3. Reconnect wrapper that retries on connection-class errors only.
 *
 * Usage:
 *   npm install
 *   export DVARA_URL="http://localhost:8080/v1"
 *   export DVARA_API_KEY="gw_<your-dvara-key>"
 *   npm start
 *
 * See https://dvarahq.com/docs/cookbook/streaming-end-to-end for the
 * narrated walkthrough.
 */

import OpenAI, { APIConnectionError } from "openai";

const client = new OpenAI({
  apiKey: process.env.DVARA_API_KEY ?? "your-dvara-api-key",
  baseURL: process.env.DVARA_URL ?? "http://localhost:8080/v1",
});

async function basicStreaming() {
  console.log("=== 1. Basic streaming ===");
  const stream = await client.chat.completions.create({
    model: "gpt-4o",
    messages: [{ role: "user", content: "Write a haiku about programming." }],
    stream: true,
  });
  for await (const chunk of stream) {
    const delta = chunk.choices[0]?.delta;
    if (delta?.content) process.stdout.write(delta.content);
    if (chunk.choices[0]?.finish_reason === "content_filter") {
      process.stdout.write("\n[stream stopped: governance enforcement]");
    }
  }
  console.log("\n");
}

async function jsonSchemaStreaming() {
  console.log("=== 2. json_schema streaming ===");
  const schema = {
    type: "object",
    properties: {
      name: { type: "string" },
      age: { type: "integer" },
    },
    required: ["name", "age"],
  };
  const stream = await client.chat.completions.create({
    model: "gpt-4o",
    messages: [{ role: "user", content: "Extract: John is 30 years old." }],
    response_format: {
      type: "json_schema",
      json_schema: { name: "person", schema, strict: true },
    },
    stream: true,
  });

  let buffer = "";
  for await (const chunk of stream) {
    const delta = chunk.choices[0]?.delta?.content;
    if (delta) buffer += delta;
  }
  const result = JSON.parse(buffer);
  console.log(`name: ${result.name}, age: ${result.age}\n`);
}

/**
 * Yield chunks with exponential-backoff retries on connection errors.
 * Does NOT retry on content_filter — that's an enforcement decision,
 * the same content would trigger the same rule on retry.
 */
async function* streamWithRetry(messages, attempts = 3, baseBackoff = 500) {
  for (let attempt = 0; attempt < attempts; attempt++) {
    try {
      const stream = await client.chat.completions.create({
        model: "gpt-4o",
        messages,
        stream: true,
      });
      for await (const chunk of stream) yield chunk;
      return;
    } catch (err) {
      if (!(err instanceof APIConnectionError) || attempt === attempts - 1) throw err;
      await new Promise((resolve) => setTimeout(resolve, baseBackoff * 2 ** attempt));
    }
  }
}

async function reconnectDemo() {
  console.log("=== 3. Reconnect wrapper ===");
  for await (const chunk of streamWithRetry([
    { role: "user", content: "Say hello briefly." },
  ])) {
    const delta = chunk.choices[0]?.delta?.content;
    if (delta) process.stdout.write(delta);
  }
  console.log("\n");
}

await basicStreaming();
await jsonSchemaStreaming();
await reconnectDemo();
